src/main/kotlin/com/craftschemers/hub/minigame/Minigame.kt

```diff
@@ -1,18 +1,71 @@
 package com.craftschemers.hub.minigame
 
+import com.craftschemers.hub.Hub
 import com.craftschemers.hub.HubPlayer
+import org.bukkit.Bukkit
+import org.bukkit.ChatColor
+import org.bukkit.Location
 
-abstract class Minigame(val name: String, val type: GameType) {
-    var state = GameState.LOBBY
-    val players = mutableListOf<HubPlayer>()
+abstract class Minigame(
+    val name: String,
+    val displayName: String,
+    val type: GameType,
+    val minPlayers: Int,
+    val maxPlayers: Int,
+    val maxServers: Int = 1,
+) {
 
-    fun addPlayer(player: HubPlayer) {
-        players += player
-        player.game = type
+    var lobbyLocation: Location? = null
+
+    init {
+        val configFile = Hub.plugin.config
+        val worldName = configFile.getString("$name.lobby.world")
+
+        val x = configFile.getDouble("$name.lobby.x")
+        val y = configFile.getDouble("$name.lobby.y")
+        val z = configFile.getDouble("$name.lobby.z")
+
+        val world = worldName?.let { Bukkit.getWorld(it) }
+
+        lobbyLocation = worldName?.let { Location(world, x, y, z) }
     }
 
-    fun hasPlayer(player: HubPlayer): Boolean = player in players
+    open val lobbies = mutableListOf<AbstractGameData>()
 
-    abstract fun start()
+    fun addPlayer(player: HubPlayer) {
+        println("attempting to add $player")
+        if (maxServers == 0) return
+        val lobby: AbstractGameData?
+        if (lobbies.size == 0) {
+            lobby = createNewLobby()
+            lobby.addPlayer(player)
+            lobbies.add(lobby)
+        } else {
+            // find first available lobby or make a new one
+            lobby = lobbies.find { it.state == GameState.LOBBY }
+            if (lobby == null && lobbies.size < maxServers) {
+                // create a new lobby
+                val newLobby = createNewLobby()
+                newLobby.players.add(player)
+                lobbies.add(newLobby)
+                return
+            } else if (lobby == null || lobby.players.size >= maxPlayers) {
+                // spectate (cant create any more lobbies)
+                TODO()
+            } else {
+                lobby.addPlayer(player)
+            }
+        }
+        for (p in lobby.players) {
```

Comment:
Could do this functionally
```suggestion
        lobby.players.forEach {
```